### PR TITLE
Allow plugins to override boot failure view

### PR DESCRIPTION
**Pull request by timja**
See [JENKINS-XXXXX](https://issues.jenkins.io/browse/JENKINS-XXXXX).

Currently it does not seem possible to provide your own `BootFailure` subclass for a startup error, Jenkins just displays a huge stacktrace with an error and you can't provide a nice error.

<details>
<summary>Before</summary>

image

</details>

<details>
<summary>After</summary>

image

</details>

### Testing done

Manually with https://github.com/jenkinsci/configuration-as-code-plugin/pull/2352

Created a file called `jenkins.yaml` in the `work` directory with these contents:
```yaml
invalid:
  systemMessage2: Hello World

```

I have not been able to figure out how to write a unit test for this, at least with JTH the process is destroyed forcefully if Jenkins does not start and there's no way to get the information.

### Proposed changelog entries

- `BootFailure` subclasses can now override the Jenkins failed to startup page

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/8442"></a>

### CODE REVIEW

**Review by NotMyFault — APPROVED** [review]

**Comment by timja** [comment]
@jenkinsci/core anyone mind taking a look please?

File: core/src/main/java/hudson/WebAppMain.java
Comment (jglick) [inline core/src/main/java/hudson/WebAppMain.java:285]: Could be simplified to return `@CheckForNull BootFailure`.

**Review by jglick — APPROVED** [review]
Logic could probably be simplified, but looks OK as it is.

**Comment by timja** [comment]
> > with JTH the process is destroyed forcefully if Jenkins does not start and there's no way to get the information
> 
> Might be a good addition to `RealJenkinsRule`: a variant of `startJenkins` that does not wait for the ping endpoint to response, letting the test code use e.g. HtmlUnit to query the Jenkins UI as it is.

maybe but the use-case seems very limited.

----

/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
